app/controllers/datacite_dois_controller.rb

```diff
@@ -778,441 +777,26 @@ def safe_params
-        p[:doi] = generate_random_dois(p[:prefix], number: 123_456).first
-      elsif p[:doi].blank? && p[:prefix].present?
-        p[:doi] = generate_random_dois(p[:prefix]).first
-      end
-
-      # replace DOI, but otherwise don't touch the XML
-      # use Array.wrap(read_attrs.first) as read_attrs may also be [[]]
-      if meta["from"] == "datacite" && Array.wrap(read_attrs.first).blank?
-        xml = replace_doi(xml, doi: p[:doi] || meta["doi"])
-      elsif xml.present? || Array.wrap(read_attrs.first).present?
-        regenerate = true
-      end
-
-      p[:xml] = xml if xml.present?
-
-
-      read_attrs_keys = %i[
-        url
-        creators
-        contributors
-        titles
-        publisher
-        publicationYear
-        types
-        descriptions
-        container
-        sizes
-        formats
-        language
-        dates
-        identifiers
-        relatedIdentifiers
-        relatedItems
-        fundingReferences
-        geoLocations
-        rightsList
-        agency
-        subjects
-        contentUrl
-        schemaVersion
-      ]
-
-      # merge attributes from xml into regular attributes
-      # make sure we don't accidentally set any attributes to nil
-      read_attrs_keys.each do |attr|
-        if p.has_key?(attr) || meta.has_key?(attr.to_s.underscore)
-          p.merge!(
-            attr.to_s.underscore =>
-              p[attr] || meta[attr.to_s.underscore] || p[attr],
-          )
-        end
-      end
-
-      # handle version metadata
-      if p.has_key?(:version) || meta["version_info"].present?
-        p[:version_info] = p[:version] || meta["version_info"]
-      end
-
-      # only update landing_page info if something is received via API to not overwrite existing data
-      p[:landing_page] = p[:landingPage] if p[:landingPage].present?
-
-      p.merge(regenerate: p[:regenerate] || regenerate).except(
-        # ignore camelCase keys, and read-only keys
-        :confirmDoi,
-        :prefix,
-        :suffix,
-        :publicationYear,
-        :alternateIdentifiers,
-        :rightsList,
-        :relatedIdentifiers,
-        :relatedItems,
-        :fundingReferences,
-        :geoLocations,
-        :metadataVersion,
-        :schemaVersion,
-        :state,
-        :mode,
-        :isActive,
-        :landingPage,
-        :created,
-        :registered,
-        :updated,
-        :published,
-        :lastLandingPage,
-        :version,
-        :lastLandingPageStatus,
-        :lastLandingPageStatusCheck,
-        :lastLandingPageStatusResult,
-        :lastLandingPageContentType,
-        :contentUrl,
-        :viewsOverTime,
-        :downloadsOverTime,
-        :citationsOverTime,
-        :citationCount,
-        :downloadCount,
-        :partCount,
-        :partOfCount,
-        :referenceCount,
-        :versionCount,
-        :versionOfCount,
-        :viewCount,
-      )
+    def sanitized_params
+      ParamsSanitizer.new(safe_params.to_h).cleanse
```

Comment:
If `safe_params` should always return a hash, the `.to_h` might be redundant

Reply:
Actually, this is necessary because of strong params. Otherwise, you will be cloning the params object. this is practically one of the reasons safe params is so big.

Reply:
Thanks - the strong params part was the bit of knowledge I was missing here